tools/transform.cpp

```diff
@@ -285,6 +285,73 @@ static expr preprocess(Transform &t, const set<expr> &qvars0,
 }
 
 
+static expr
+encode_undef_refinement_per_elem(
+    const Type &ty, const expr &a, const expr &a2, const expr &a_np,
+    const expr &b, const expr &b2) {
+  const auto *aty = ty.getAsAggregateType();
+  if (!aty)
+    return (a_np && a == a2).implies(b == b2);
+
+  StateValue sva{expr(a), expr(a_np)}, sva2{expr(a2), expr()};
+  StateValue svb{expr(b), expr()}, svb2{expr(b2), expr()};
+  expr result = true;
+
+  for (unsigned i = 0; i < aty->numElementsConst(); ++i) {
+    if (aty->isPadding(i))
+      continue;
+
+    result &= encode_undef_refinement_per_elem(aty->getChild(i),
+        aty->extract(sva, i).value, aty->extract(sva2, i).value,
+        aty->extract(sva, i).non_poison, aty->extract(svb, i).value,
+        aty->extract(svb2, i).value);
+  }
+  return result;
+}
+
+static expr
+encode_undef_refinement(const Type &type, const State::ValTy &ap,
+                        const State::ValTy &bp, set<expr> &qvars) {
```

Comment:
I think we want to pass qvars by value here so the new qvars don't leak to the remaining queries

Reply:
This is intended because source's new undef vars should be added to qvars.

Reply:
ok, but then you are relying on Z3's preprocessor to remove all those useless quantified vars for the other queries. Let's see the benchmarks..